src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -314,17 +322,35 @@ private void AssertIfNull(ImmutableArray<DiagnosticData> diagnostics)
         }
 
         [Conditional("DEBUG")]
-        private void AssertIfNull(DiagnosticData diagnostic)
+        private void AssertIfNull<T>(T obj) where T : class
         {
-            if (diagnostic == null)
+            if (obj == null)
             {
                 Contract.Requires(false, "who returns invalid data?");
             }
         }
 
+        [Conditional("DEBUG")]
+        public void AssertCleanup()
+        {
+            foreach (var idAndData in _map.Values.ToList())
+            {
+                foreach (var data in idAndData.Values.ToList())
+                {
+                    AssertIfNull(data.Workspace);
+                }
+            }
+        }
+
         private struct Data : IEquatable<Data>
         {
-            public readonly Workspace Workspace;
+            private static readonly ConditionalWeakTable<Workspace, WeakReference<Workspace>> s_cache =
```

Comment:
> why create so many redundant weak reference objects?

I would prefer we not create weak references in the first place.  If you are goin gto make them, the last thing i want is then a CWT wrapper around that that means we have weak references to weak references.  INtroducing that much complexity to save 300 allocations (in a systme of literally billions of allocations) seems unnecessary.  It's is a micro optimization that makes the code much less clear while adding no appreciable value.

--

Before going on with more discusison on the WeakReference stuff, can you please explain hte lifetimes of these objects and why it's necessary that we have to work with Data objects that live longer than the workspace entities they were created from?  This seems extremely strange and i'm trying to find out why that's necessary.

If it is necessary, i'd like an approach where we don't hold onto Workspaces at all (either directly, or through weak references). We should store something else (like an ID) so that the objects stay small, and are truly data, and not things that can hold onto massive graphs of objects.